### DISCUSSION TITLE
Use of Parameters

### DISCUSSION THREAD

**Discussion by pmelchior** [pmelchior/scarlet2#55] [Polls]
We currently have a workable approach to the question which parameters in a model we like to optimize or sample from. 
1. Any `scarlet2.Module` that receives a `Parameter` instance will assume that is can vary it for optimization/sampling.
2. If it receives another numerical type of array thereof, it assumes that it's fixed.

In case 1, the `Parameter` instance will immediately (that is at `__init__`) be converted to its base value and information (step sizes, priors) put into a dictionary. The advantage is that we never have to deal with `Parameters` anywhere in the code, it's always `jnp.arrays` or regular numerical types. This allows `scarlet2` models to be evaluated by other optimization or sampling frameworks (one could run it with `emcee` or Numpyro, etc).

But it also creates overhead because we have to create and pass around these dictionaries, make sure that the parameters and their dictionaries are consistent, etc. The step of conversion during `__init__` (actually during `__post_init__` if you care about dataclass semantics) cannot be enforced for derived classes, so it relies on the coder to ensure that it's actually called. That's dangerous. Newer versions of `equinox` spit out warnings about this coding pattern.

For us, this is practically only ever relevant for `Scene` because it's the only place where we provide fitting and sampling routines. Again, a user could use their own, but it's important to have working implementation for either as part of the library.

So, I'd like to know if people prefer the current option or an alternative:

## Option 1 (current)

The module initialization distinguishes between (fixed) arrays and (non-fixed) `Parameters`:

```python
with Scene(model_frame) as scene:
        Source(
            center, # not optimized/sampled
           ArraySpectrum(Parameter(spectrum, constraint=constraints.positive, stepsize=relative_step)),
           ArrayMorphology(Parameter(morph, prior=galaxygrad.hsc_prior, stepsize=0.01))
        )

scene_ = scene.fit(obs, maxiter=200)
```

Note that `scene.fit` knows what it needs to optimize because it has been declared before.

## Option 2 (alternative)

The initialization works with simple numerical types:

```python
with Scene(model_frame) as scene:
        Source(
            center,
           ArraySpectrum(spectrum),
           ArrayMorphology(morph)
        )

parameters = dict(
    "sources.0.spectrum": dict("constraint"=constraints.positive, "stepsize"=relative_step),
    "sources.0.morphology": dict("prior"=galaxygrad.hsc_prior, "stepsize"=0.01)
}
scene_ = scene.fit(obs, parameters, maxiter=200)
```
The main difference is that the decision what to fit/sample will be made _after_ creating a model. Which means its avoid having to create these info dictionaries at init time. It makes the model definition independent of the fitting. That also means, if we want to run another fitter/sampling thereafter changing other parameters, we can modify the `parameters` and call `fit` again.

We can tinker with the best way of providing the infos about the parameters to the fitter/sampler, e.g.

```python
with Scene() as scene:
    ...

parameters = 
    Param("sources.0.spectrum", constraint=constraints.positive, stepsize=relative_step),
    Param("sources.0.morphology", prior=galaxygrad.hsc_prior, stepsize=0.01)

scene_ = scene.fit(obs, parameters, maxiter=200)
```

This is more concise than before but basically the same idea. Btw, this lookup by name is what we already use (it makes the lookup independent of the specific instance, so it works for `scene` and for `scene_`), the difference is that we don't store the result inside of `Module`, we keep it in user space.

**Comment by b-remy** [comment]
I like option 2 which gives us more control on what parameters we want to fit or sample

**Comment by charlotteaward** [comment]
I also think I think I like option 2 because it will be clearer what is fixed or not by default compared to the previous approach when I was using `scene.set_info('sources.0.morphology.center', fixed=False,stepsize = pos_step)` etc
